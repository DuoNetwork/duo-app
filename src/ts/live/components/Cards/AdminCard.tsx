import { Table } from 'antd';
import * as d3 from 'd3';
import * as React from 'react';
import * as CST from '../../common/constants';
import contractUtil from '../../common/contractUtil';
import { IAddresses, ICustodianStates } from '../../common/types';
import AdminInputButton from '../Common/AdminInputButton';
import { SCard, SCardTitle, STableWrapper } from './_styled';

const { Column } = Table;

interface IProps {
	addresses: IAddresses;
	states: ICustodianStates;
	account: string;
}

export default class AdminCard extends React.PureComponent<IProps> {
	public render() {
		const { states, account, addresses } = this.props;
		const data: object[] = [];
		data.push({
			key: 0,
			[CST.TH_STATE]: 'Fee Accumulated',
			[CST.TH_VALUE]: d3.format(',.2f')(states.feeAccumulated),
			[CST.TH_ACTION]: (
				<AdminInputButton
					account={account}
					type={CST.TH_COLLLECT_FEE}
					disabled={account === '0x0' || addresses.feeCollector.address !== account}
					validateInput={(value: string) =>
						!value || (value.match(CST.RX_NUM_P) && Number(value) > 0)
							? ''
							: 'Invalid number'
					}
				/>
			)
		});
		data.push({
			key: 1,
			[CST.TH_STATE]: 'ETH:DUO Fee Ratio',
			[CST.TH_VALUE]: states.ethDuoFeeRatio,
			[CST.TH_ACTION]: (
				<AdminInputButton
					account={account}
					type={CST.TH_SET_VALUE}
					disabled={account === '0x0' || addresses.operator.address !== account}
					index={1}
					validateInput={(value: string) =>
						!value || (value.match(CST.RX_INTEGER) && Number(value) > 0)
							? ''
							: 'Invalid number'
					}
				/>
			)
		});
		data.push({
			key: 2,
			[CST.TH_STATE]: 'Iteration Gas Threshold',
			[CST.TH_VALUE]: states.iterationGasThreshold,
			[CST.TH_ACTION]: (
				<AdminInputButton
					account={account}
					type={CST.TH_SET_VALUE}
					disabled={account === '0x0' || addresses.operator.address !== account}
					index={2}
					validateInput={(value: string) =>
						!value || (value.match(CST.RX_INTEGER) && Number(value) > 0)
							? ''
							: 'Invalid number'
					}
				/>
			)
		});
		data.push({
			key: 3,
			[CST.TH_STATE]: 'PreReset Waiting Blocks',
			[CST.TH_VALUE]: states.preResetWaitingBlocks,
			[CST.TH_ACTION]: (
				<AdminInputButton
					account={account}
					type={CST.TH_SET_VALUE}
					disabled={account === '0x0' || addresses.operator.address !== account}
					index={3}
					validateInput={(value: string) =>
						!value || (value.match(CST.RX_INTEGER) && Number(value) > 0)
							? ''
							: 'Invalid number'
					}
				/>
			)
		});
		data.push({
			key: 4,
			[CST.TH_STATE]: 'Price Tolerance',
			[CST.TH_VALUE]: states.priceTol,
			[CST.TH_ACTION]: (
				<AdminInputButton
					account={account}
					type={CST.TH_SET_VALUE}
					disabled={account === '0x0' || addresses.operator.address !== account}
					index={4}
					validateInput={(value: string) =>
						!value || (value.match(CST.RX_NUM_P) && Number(value) > 0)
							? ''
							: 'Invalid number'
					}
				/>
			)
		});
		data.push({
			key: 5,
			[CST.TH_STATE]: 'Price Feed Tolerance',
			[CST.TH_VALUE]: states.priceFeedTol,
			[CST.TH_ACTION]: (
				<AdminInputButton
					account={account}
					type={CST.TH_SET_VALUE}
					disabled={account === '0x0' || addresses.operator.address !== account}
					index={5}
					validateInput={(value: string) =>
						!value || (value.match(CST.RX_NUM_P) && Number(value) > 0)
							? ''
							: 'Invalid number'
					}
				/>
			)
		});
		data.push({
			key: 6,
			[CST.TH_STATE]: 'Price Feed Time Tolerance',
			[CST.TH_VALUE]: states.priceFeedTimeTol,
			[CST.TH_ACTION]: (
				<AdminInputButton
					account={account}
					type={CST.TH_SET_VALUE}
					disabled={account === '0x0' || addresses.operator.address !== account}
					index={6}
					validateInput={(value: string) =>
						!value || (value.match(CST.RX_INTEGER) && Number(value) > 0)
							? ''
							: 'Invalid number'
					}
				/>
			)
		});
		data.push({
			key: 7,
			[CST.TH_STATE]: 'Price Update Cool Down',
			[CST.TH_VALUE]: states.priceUpdateCoolDown,
			[CST.TH_ACTION]: (
				<AdminInputButton
					account={account}
					type={CST.TH_SET_VALUE}
					disabled={account === '0x0' || addresses.operator.address !== account}
					index={7}
					validateInput={(value: string) =>
						!value || (value.match(CST.RX_INTEGER) && Number(value) > 0)
							? ''
							: 'Invalid number'
					}
				/>
			)
		});
		data.push({
			key: 8,
			[CST.TH_STATE]: 'Address Pool Length',
			[CST.TH_VALUE]: states.addrPoolLength,
			[CST.TH_ACTION]: (
				<AdminInputButton
					account={account}
					type={CST.TH_ADD_ADDR}
					disabled={account === '0x0' || addresses.poolManager.address !== account}
					validateInput={(value: string) =>
						contractUtil.checkAddress(value) ? '' : 'Invalid Address'
					}
				/>
			)
		});
		data.push({
			key: 9,
			[CST.TH_STATE]: 'Staged Prices',
			[CST.TH_VALUE]: states.numOfPrices
		});
		data.push({
			key: 10,
			[CST.TH_STATE]: 'Next Reset Addr Index',
			[CST.TH_VALUE]: states.nextResetAddrIndex
		});
		data.push({
			key: 11,
			[CST.TH_STATE]: 'Last Admin Time',
			[CST.TH_VALUE]: states.lastAdminTime
		});
		data.push({
			key: 12,
			[CST.TH_STATE]: 'Admin Cool Down',
			[CST.TH_VALUE]: states.adminCoolDown
		});
		return (
			<SCard
				title={<SCardTitle>{CST.TH_ADMIN.toUpperCase()}</SCardTitle>}
				width="1000px"
				margin="0 0px 0 0"
				inlinetype="table"
			>
				<STableWrapper>
					<Table
						dataSource={data}
						pagination={false}
						rowClassName={() => 'admin-table-row'}
					>
						{[CST.TH_STATE, CST.TH_VALUE, CST.TH_ACTION].map(th => (
							<Column title={th} dataIndex={th} key={th} />
						))}
					</Table>
				</STableWrapper>
			</SCard>
		);
	}
}
