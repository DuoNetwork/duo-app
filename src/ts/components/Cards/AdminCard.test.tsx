// fix for @ledgerhq/hw-transport-u2f 4.28.0
import '@babel/polyfill';
import { Constants as WrapperConstants } from '@finbook/duo-contract-wrapper';
import { shallow } from 'enzyme';
import * as React from 'react';
import AdminCard from './AdminCard';

describe('AuthCard Test', () => {
	describe('User Login', () => {
		const addresses = {
			feeCollector: '0x0017d61f0B0a28E2F0eBB3B6E269738a6252CFeD',
			oracle: '0x00D8d0660b243452fC2f996A892D3083A903576F',
			aToken: '0x00184D7745ef135490114EEFfB762C2A60E067d3',
			bToken: '0x0022BFd6AFaD3408A1714fa8F9371ad5Ce8A0F1a',
			operator: '0x0022BFd6AFaD3408A1714fa8F9371ad5Ce8A0F1a',
			roleManager: '0x0022BFd6AFaD3408A1714fa8F9371ad5Ce8A0F1a'
		};

		const account = '0x0';
		// const account1 = '0x01';

		const states = {
			resetState: 'test',
			alpha: 1,
			beta: 1,
			periodCoupon: 0.000017,
			limitPeriodic: 1.035,
			limitLower: 0.25,
			limitUpper: 2,
			iterationGasThreshold: 65000,
			state: 'Trading',
			minBalance: 1,
			totalSupplyA: 82880.4753362674,
			totalSupplyB: 82880.4753362674,
			collateral: 1,
			navA: 1.026299,
			navB: 0.5862017153842389,
			lastPrice: 12,
			lastPriceTime: 12,
			resetPrice: 12,
			resetPriceTime: 12,
			createCommRate: 0.01,
			redeemCommRate: 0.01,
			period: 3600,
			maturity: 12,
			preResetWaitingBlocks: 10,
			priceFetchCoolDown: 123,
			nextResetAddrIndex: 0,
			totalUsers: 123,
			feeBalance: 123,
			lastOperationTime: 123,
			operationCoolDown: 123
		};

		it('Test Snapshot', () => {
			const wrapper1 = shallow(
				<AdminCard
					type={WrapperConstants.BEETHOVEN}
					tenor={WrapperConstants.TENOR_PPT}
					addresses={addresses}
					states={states}
					account={account}
				/>
			);
			expect(wrapper1).toMatchSnapshot();
			const wrapper2 = shallow(
				<AdminCard
					type={WrapperConstants.BEETHOVEN}
					tenor={WrapperConstants.TENOR_PPT}
					addresses={addresses}
					states={states}
					account={'0x01'}
				/>
			);
			expect(wrapper2).toMatchSnapshot();
		});
		// test('dummy', () => expect(true).toBeTruthy());
	});
});
