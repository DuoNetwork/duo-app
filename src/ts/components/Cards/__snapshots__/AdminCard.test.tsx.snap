// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthCard Test User Login Test Snapshot 1`] = `
<Styled(Card)
  inlinetype="table"
  margin="0 0px 0 0"
  title={
    <styled.div>
      ADMIN
    </styled.div>
  }
  width="1000px"
>
  <styled.div>
    <Table
      bordered={false}
      className=""
      dataSource={
        Array [
          Object {
            "Action": <AdminInputButton
              account="0x0"
              disabled={true}
              tenor="test"
              type="Collect Fee"
              validateInput={[Function]}
            />,
            "State": "Fee Accumulated",
            "Value": "123.00",
            "key": 0,
          },
          Object {
            "Action": <AdminInputButton
              account="0x0"
              disabled={true}
              index={0}
              tenor="test"
              type="Set Value"
              validateInput={[Function]}
            />,
            "State": "Create Comm Rate BP",
            "Value": 100,
            "key": 1,
          },
          Object {
            "Action": <AdminInputButton
              account="0x0"
              disabled={true}
              index={1}
              tenor="test"
              type="Set Value"
              validateInput={[Function]}
            />,
            "State": "Redeem Comm Rate BP",
            "Value": 100,
            "key": 2,
          },
          Object {
            "Action": <AdminInputButton
              account="0x0"
              disabled={true}
              index={2}
              tenor="test"
              type="Set Value"
              validateInput={[Function]}
            />,
            "State": "Iteration Gas Threshold",
            "Value": 65000,
            "key": 3,
          },
          Object {
            "Action": <AdminInputButton
              account="0x0"
              disabled={true}
              index={3}
              tenor="test"
              type="Set Value"
              validateInput={[Function]}
            />,
            "State": "PreReset Waiting Blocks",
            "Value": 10,
            "key": 4,
          },
          Object {
            "State": "Next Reset Addr Index",
            "Value": 0,
            "key": 5,
          },
          Object {
            "State": "Last Operation Time",
            "Value": 123,
            "key": 6,
          },
          Object {
            "State": "Operation Cool Down",
            "Value": 123,
            "key": 7,
          },
        ]
      }
      indentSize={20}
      loading={false}
      locale={Object {}}
      pagination={false}
      prefixCls="ant-table"
      rowClassName={[Function]}
      rowKey="key"
      showHeader={true}
      size="default"
      useFixedHeader={false}
    >
      <Column
        dataIndex="State"
        key="State"
        title="State"
      />
      <Column
        dataIndex="Value"
        key="Value"
        title="Value"
      />
      <Column
        dataIndex="Action"
        key="Action"
        title="Action"
      />
    </Table>
  </styled.div>
</Styled(Card)>
`;

exports[`AuthCard Test User Login Test Snapshot 2`] = `
<Styled(Card)
  inlinetype="table"
  margin="0 0px 0 0"
  title={
    <styled.div>
      ADMIN
    </styled.div>
  }
  width="1000px"
>
  <styled.div>
    <Table
      bordered={false}
      className=""
      dataSource={
        Array [
          Object {
            "Action": <AdminInputButton
              account="0x01"
              disabled={true}
              tenor="test"
              type="Collect Fee"
              validateInput={[Function]}
            />,
            "State": "Fee Accumulated",
            "Value": "123.00",
            "key": 0,
          },
          Object {
            "Action": <AdminInputButton
              account="0x01"
              disabled={true}
              index={0}
              tenor="test"
              type="Set Value"
              validateInput={[Function]}
            />,
            "State": "Create Comm Rate BP",
            "Value": 100,
            "key": 1,
          },
          Object {
            "Action": <AdminInputButton
              account="0x01"
              disabled={true}
              index={1}
              tenor="test"
              type="Set Value"
              validateInput={[Function]}
            />,
            "State": "Redeem Comm Rate BP",
            "Value": 100,
            "key": 2,
          },
          Object {
            "Action": <AdminInputButton
              account="0x01"
              disabled={true}
              index={2}
              tenor="test"
              type="Set Value"
              validateInput={[Function]}
            />,
            "State": "Iteration Gas Threshold",
            "Value": 65000,
            "key": 3,
          },
          Object {
            "Action": <AdminInputButton
              account="0x01"
              disabled={true}
              index={3}
              tenor="test"
              type="Set Value"
              validateInput={[Function]}
            />,
            "State": "PreReset Waiting Blocks",
            "Value": 10,
            "key": 4,
          },
          Object {
            "State": "Next Reset Addr Index",
            "Value": 0,
            "key": 5,
          },
          Object {
            "State": "Last Operation Time",
            "Value": 123,
            "key": 6,
          },
          Object {
            "State": "Operation Cool Down",
            "Value": 123,
            "key": 7,
          },
        ]
      }
      indentSize={20}
      loading={false}
      locale={Object {}}
      pagination={false}
      prefixCls="ant-table"
      rowClassName={[Function]}
      rowKey="key"
      showHeader={true}
      size="default"
      useFixedHeader={false}
    >
      <Column
        dataIndex="State"
        key="State"
        title="State"
      />
      <Column
        dataIndex="Value"
        key="Value"
        title="Value"
      />
      <Column
        dataIndex="Action"
        key="Action"
        title="Action"
      />
    </Table>
  </styled.div>
</Styled(Card)>
`;
